### ISSUE TITLE
Admin Login Unusable on PHP 7.4 Due to PHP Notices

### ISSUE THREAD

**Issue by davidalger** [magento/magento2#28729]
### Preconditions (*)

1. Magento marketplace keys with access to 2.4 alphas (to mimic my testing, which is installing from composer, not the git repo)
2. PHP Notice reporting is enabled (as it is by default in `app/bootstrap.php`)
3. Target environment is running PHP 7.4.x (my testing was running PHP 7.4.6)

### Steps to reproduce (*)

1. Install Magento 2.4.0-alpha9 on a web stack running PHP 7.4
2. Open the admin login page in a web browser
3. Observe the PHP Notice preventing the admin login page from displaying

Alternately, to completely match the test environment used by me:

1. Clone [this project template](https://github.com/davidalger/warden-env-magento2)
2. Edit the `.env` file so the environment will use PHP 7.4 images
3. Install Magento 2.4.0 alpha via `./tools/init.sh --clean-install --meta-version 2.4.0-alpha9`
4. Observe the PHP Notice preventing the admin login page from displaying

### Expected result (*)

1. A **working** Magento admin login page: https://dropshare-ot3kdw.s3.amazonaws.com/uLum9y/Screen-Shot-2020-06-15-10-25-47.66-QYOpfYFpN6Zm.png

### Actual result (*)

1. A **broken** Magento admin login page: https://dropshare-ot3kdw.s3.amazonaws.com/uLum9y/Screen-Shot-2020-06-15-10-26-00.90-hpKrZnimxhx6.png

```
1 exception(s):
Exception #0 (Exception): Notice: Trying to access array offset on value of type null in /var/www/html/vendor/magento/framework/View/Element/Html/Calendar.php on line 112

Exception #0 (Exception): Notice: Trying to access array offset on value of type null in /var/www/html/vendor/magento/framework/View/Element/Html/Calendar.php on line 112
<pre>#1 Magento\Framework\View\Element\Html\Calendar->_toHtml() called at [vendor/magento/framework/View/Element/AbstractBlock.php:1111]
#2 Magento\Framework\View\Element\AbstractBlock->Magento\Framework\View\Element\{closure}() called at [vendor/magento/framework/View/Element/AbstractBlock.php:1115]
#3 Magento\Framework\View\Element\AbstractBlock->_loadCache() called at [vendor/magento/framework/View/Element/AbstractBlock.php:675]
#4 Magento\Framework\View\Element\AbstractBlock->toHtml() called at [vendor/magento/framework/View/Layout.php:566]
#5 Magento\Framework\View\Layout->_renderBlock() called at [vendor/magento/framework/View/Layout.php:542]
#6 Magento\Framework\View\Layout->renderNonCachedElement() called at [generated/code/Magento/Framework/View/Layout/Interceptor.php:206]
#7 Magento\Framework\View\Layout\Interceptor->renderNonCachedElement() called at [vendor/magento/framework/View/Layout.php:497]
#8 Magento\Framework\View\Layout->renderElement() called at [generated/code/Magento/Framework/View/Layout/Interceptor.php:193]
#9 Magento\Framework\View\Layout\Interceptor->renderElement() called at [vendor/magento/framework/View/Layout.php:594]
#10 Magento\Framework\View\Layout->_renderContainer() called at [vendor/magento/framework/View/Layout.php:544]
#11 Magento\Framework\View\Layout->renderNonCachedElement() called at [generated/code/Magento/Framework/View/Layout/Interceptor.php:206]
#12 Magento\Framework\View\Layout\Interceptor->renderNonCachedElement() called at [vendor/magento/framework/View/Layout.php:497]
#13 Magento\Framework\View\Layout->renderElement() called at [generated/code/Magento/Framework/View/Layout/Interceptor.php:193]
#14 Magento\Framework\View\Layout\Interceptor->renderElement() called at [vendor/magento/framework/View/Layout.php:594]
#15 Magento\Framework\View\Layout->_renderContainer() called at [vendor/magento/framework/View/Layout.php:544]
#16 Magento\Framework\View\Layout->renderNonCachedElement() called at [generated/code/Magento/Framework/View/Layout/Interceptor.php:206]
#17 Magento\Framework\View\Layout\Interceptor->renderNonCachedElement() called at [vendor/magento/framework/View/Layout.php:497]
#18 Magento\Framework\View\Layout->renderElement() called at [generated/code/Magento/Framework/View/Layout/Interceptor.php:193]
#19 Magento\Framework\View\Layout\Interceptor->renderElement() called at [vendor/magento/framework/View/Layout.php:963]
#20 Magento\Framework\View\Layout->getOutput() called at [generated/code/Magento/Framework/View/Layout/Interceptor.php:492]
#21 Magento\Framework\View\Layout\Interceptor->getOutput() called at [vendor/magento/framework/View/Result/Page.php:258]
#22 Magento\Framework\View\Result\Page->render() called at [vendor/magento/framework/View/Result/Layout.php:171]
#23 Magento\Framework\View\Result\Layout->renderResult() called at [generated/code/Magento/Backend/Model/View/Result/Page/Interceptor.php:193]
#24 Magento\Backend\Model\View\Result\Page\Interceptor->renderResult() called at [vendor/magento/framework/App/Http.php:120]
#25 Magento\Framework\App\Http->launch() called at [generated/code/Magento/Framework/App/Http/Interceptor.php:24]
#26 Magento\Framework\App\Http\Interceptor->launch() called at [vendor/magento/framework/App/Bootstrap.php:263]
#27 Magento\Framework\App\Bootstrap->run() called at [pub/index.php:40]
</pre>
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### Workaround

1. Employ PHP 7.3 rather than PHP 7.4; not a great solution.

**Comment by mrtuvn** [comment]
@davidalger this class also recall in https://github.com/magento/magento2/issues/28081 by @lbajsarowicz 
Seem a quite lot of files got error in php7.4

**Comment by lbajsarowicz** [comment]
Ouch. That was reported during PHP 7.4 migration to Core Team, that we need to stabilize such.
I'll check what we can do now with these.

**Comment by davidalger** [comment]
cc// @magento/magento-beta

**Comment by davidalger** [comment]
Update: tried this on the `2.4.0-beta1` and the issue is present there as well

**Comment by fascinosum** [comment]
@davidalger, thank you for you contribution. Could you help narrow down the  issue? Were you able to reproduce it with any other development environment?
You can even [request a test environment](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#vanilla-pr) here
Please review the services you use, they should meet [system requirements](https://devdocs.magento.com/guides/v2.4/install-gde/system-requirements-tech.html )
It seems that for some reason you don't have all the necessary data for the locale ResourceBundle. To be specific, for `$localeData['fields']['day']['relative']['0']` in /var/www/html/vendor/magento/framework/View/Element/Html/Calendar.php:112

**Comment by davidalger** [comment]
@fascinosum I have not tried any other dev environments, and no I did not ask for a test environment since to my knowledge I cannot specify the PHP version to use when requesting one.

> Please review the services you use, they should meet system requirements

The system I'm running meets all the requirements, and is based on the same base OS and packages I've been running Magento 2 on since it's inception. The reason I mentioned the specific dev environment used is _because_ it is container based and _because_ that means it's a 100% reproducable test case. Apart from the PHP 7.4 version difference, the environment is identical to fully functional systems running 2.3 in production.

I spent some time doing some further digging here to document some of my observations. It is perhaps the case that there is a bug here in the Magento core which has largely gone unnoticed due to PHP 7.3 failing to issue show-stopping notices as it should be and differences in `libicu` which Magento core does not currently appear to account for. The following is an isolated example of the bit of logic causing the PHP Notice on PHP 7.4, but as you'll see while the output is the same `NULL` value (indicating failure) a notice is issued on PHP 7.4 where PHP 7.3 simply masked the underlying problem:

```
$ docker run --rm -it davidalger/php:7.3 php -r '$bundle = new \ResourceBundle("en", "ICUDATA"); var_dump($bundle["fields"]["day"]["relative"]["0"]);'
NULL
```

```
$ docker run --rm -it davidalger/php:7.4 php -r '$bundle = new \ResourceBundle("en", "ICUDATA"); var_dump($bundle["fields"]["day"]["relative"]["0"]);'
PHP Notice:  Trying to access array offset on value of type null in Command line code on line 1
PHP Notice:  Trying to access array offset on value of type null in Command line code on line 1
PHP Notice:  Trying to access array offset on value of type null in Command line code on line 1
NULL
```

The images in the aforementioned environment are based on CentOS 7.8 and running the IUS php packages. The 7.3 vs 7.4 environment is identical apart from the PHP version. The following demonstrates this, identical versions of `libicu` and `intl` extension being present, with the PHP version being what differs:

```
$ docker run --rm -it davidalger/php:7.3 bash -c 'cat /etc/centos-release; yum list installed | grep libicu; echo; php -v; php --ri intl'     
CentOS Linux release 7.8.2003 (Core)
libicu.x86_64                               50.2-4.el7_7               @updates 

PHP 7.3.18 (cli) (built: Jun  4 2020 21:31:22) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.18, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.18, Copyright (c) 1999-2018, by Zend Technologies

intl

Internationalization support => enabled
ICU version => 50.2
ICU Data version => 50.2
ICU TZData version => 2018e
ICU Unicode version => 6.2

Directive => Local Value => Master Value
intl.default_locale => no value => no value
intl.error_level => 0 => 0
intl.use_exceptions => 0 => 0
```

```
$ docker run --rm -it davidalger/php:7.4 bash -c 'cat /etc/centos-release; yum list installed | grep libicu; echo; php -v; php --ri intl'
CentOS Linux release 7.8.2003 (Core)
libicu.x86_64                           50.2-4.el7_7               @updates     

PHP 7.4.7 (cli) (built: Jul  2 2020 16:10:46) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.7, Copyright (c), by Zend Technologies

intl

Internationalization support => enabled
ICU version => 50.2
ICU Data version => 50.2
ICU TZData version => 2018e
ICU Unicode version => 6.2

Directive => Local Value => Master Value
intl.default_locale => no value => no value
intl.error_level => 0 => 0
intl.use_exceptions => 0 => 0
```

Magento supports the following operating systems (and I have been running Magento 2 on EL 7.x for years):

> Linux distributions, such as RedHat Enterprise Linux (RHEL), CentOS, Ubuntu, Debian, and similar. 

CentOS 7 currently ships `libicu` version 50.2, which is certainly expected to work with Magento 2 based on the following comment (taken from the very file the PHP Notice is occurring in):

```
        /**
         * Month names in abbreviated format values was added to ICU Data tables
         * starting ICU library version 52.1. For some OS, like CentOS, default
         * installation version of ICU library is 50.1.2, which not contain
         * 'abbreviated' key, and that may cause a PHP fatal error when passing
         * as an argument of function 'iterator_to_array'. This issue affects
         * locales like ja_JP, ko_KR etc.
         *
         * @see http://source.icu-project.org/repos/icu/tags/release-50-1-2/icu4c/source/data/locales/ja.txt
         * @see http://source.icu-project.org/repos/icu/tags/release-52-1/icu4c/source/data/locales/ja.txt
         * @var \ResourceBundle $monthsData
         */
```

This comment is then followed by conditional logic to account for the abbreviated flag not being present in older versions of the ICU data. This same conditional logic it seems is needed here, as a bit of further exploration reveals that this is indeed related to another difference in ICU data format, and current code relying on `fields` as a top-level key, when it is not always available (cursory review of ICU release notes indicates it may have been added as a valid field in `libicu` 52.2). 

I'll demonstrate this below by comparing the PHP 7.3 on CentOS 7 and PHP 7.3 on Fedora 31 (which ships with a more recent `libicu` version of 63.2); output will be in attached txt files, or one may simply run the docker run commands themselves to achieve the same results:

```
$ docker run --rm -it davidalger/php:7.3 bash -c '
  cat /etc/system-release;
  yum list installed | grep libicu;
  echo;
  php -v;
  php --ri intl;
  php -r '\''
    function object_to_array($obj) {
        if (is_object($obj) || is_array($obj)) {
          $ret = [];
          foreach ($obj as $k => $v) {
            $ret[$k] = object_to_array($v);
          }
          return $ret;
        } else {
          return $obj;
        }
    }
    $bundle = new \ResourceBundle("en", "ICUDATA");
    var_dump(object_to_array($bundle));
  '\''
' > icu-50.2-data.txt
```
[icu-50.2-data.txt](https://github.com/magento/magento2/files/4882036/icu-50.2-data.txt)


```
$ docker run --rm -it fedora:31 bash -c '
  dnf install -y -q php php-intl;
  cat /etc/system-release;
  dnf list installed | grep libicu;
  echo;
  php -v;
  php --ri intl;
  php -r '\''
    function object_to_array($obj) {
        if (is_object($obj) || is_array($obj)) {
          $ret = [];
          foreach ($obj as $k => $v) {
            $ret[$k] = object_to_array($v);
          }
          return $ret;
        } else {
          return $obj;
        }
    }
    $bundle = new \ResourceBundle("en", "ICUDATA");
    var_dump(object_to_array($bundle));
  '\''
' > icu-63.2-data.txt
```
[icu-63.2-data.txt](https://github.com/magento/magento2/files/4882038/icu-63.2-data.txt)


All-in-all it would seem Magento should fall back on calendar->gregorian->fields when fields does not exist as a top-level value (with the result being normalized via strtolower):

```
$ docker run --rm -it davidalger/php:7.3 bash -c '
>   php --ri intl; php -r '\''
>     $bundle = new \ResourceBundle("en", "ICUDATA");
>     var_dump($bundle["fields"]["day"]["relative"]["0"]);
>     var_dump($bundle["calendar"]["gregorian"]["fields"]["day"]["relative"]["0"]);
>   '\''
> '

intl

Internationalization support => enabled
ICU version => 50.2
ICU Data version => 50.2
ICU TZData version => 2018e
ICU Unicode version => 6.2

Directive => Local Value => Master Value
intl.default_locale => no value => no value
intl.error_level => 0 => 0
intl.use_exceptions => 0 => 0
NULL
string(5) "Today"
```

```
$ docker run --rm -it fedora:31 bash -c '
>   dnf install -y -q php php-intl;
>   php --ri intl; php -r '\''
>     $bundle = new \ResourceBundle("en", "ICUDATA");
>     var_dump($bundle["fields"]["day"]["relative"]["0"]);
>   '\''
> '

intl

Internationalization support => enabled
ICU version => 63.2
ICU Data version => 63.2
ICU TZData version => 2019a
ICU Unicode version => 11.0

Directive => Local Value => Master Value
intl.default_locale => no value => no value
intl.error_level => 0 => 0
intl.use_exceptions => 0 => 0
string(5) "today"

```

It's worth noting that on newer versions of `libicu` the calendar->gregorian->fields does not exist, only fields as a top-level value does. So there must be conditional logic. It would not work to simply change the path to the value being sought.

I don't have time to implement (and write tests around) the solution here myself, but as 2.4 will be supporting EL as a runtime OS it would be reasonable to expect the 2.4.0 GA release to function with the version of `libicu` that ships on EL 7.

**Comment by fascinosum** [comment]
@davidalger thank you for this in-depth investigation. You are absolutely right that the current implementation in some places is based on the assumption that we have ICU library with version >= 51.x (or even >= 52.x). `fields` value in the current position has been added to ICU 51.1 (Release Date: 2013-03-22).
Due to the fact that we don't use these block variables (`today` and `week`) and never used them for any Magento 2 release, we decided not to add additional handling for instances with ICU library version < 52.x. This means we will not initialize them. You are right, it never worked properly with CentOS 7, but no one seems to really use these variables.
Ideally, we would completely remove them from the block code, but we can't do that in beta.
Internal ticket (MC-35649) has been created to track this issue

**Comment by davidalger** [comment]
> Due to the fact that we don't use these block variables (today and week) and never used them for any Magento 2 release, we decided not to add additional handling for instances with ICU library version < 51.x. 

@fascinosum Sounds good to me! As long as the notice doesn't happen, I"ll be happy. Thanks. :)

**Comment by davidalger** [comment]
The `2.4.0-beta2` release resolves this issue. Doesn't look like the commit is published, but the change was as follows:

```
--- /var/www/html/vendor/magento/framework/View/Element/Html/Calendar.php	2020-07-14 16:52:23.625564806 +0000
+++ /var/www/html/vendor/magento/framework/View/Element/Html/Calendar.php	2020-07-13 15:44:44.000000000 +0000
@@ -107,9 +108,7 @@
             ]
         );
 
-        // get "today" and "week" words
-        $this->assign('today', $this->encoder->encode($localeData['fields']['day']['relative']['0']));
-        $this->assign('week', $this->encoder->encode($localeData['fields']['week']['dn']));
+        $this->assignFieldsValues($localeData);
 
         // get "am" & "pm" words
         $this->assign('am', $this->encoder->encode($localeData['calendar']['gregorian']['AmPmMarkers']['0']));
@@ -189,5 +188,25 @@
         return (new \DateTime())->modify('- 100 years')->format('Y')
             . ':' . (new \DateTime())->modify('+ 100 years')->format('Y');
     }
-}
 
+    /**
+     * Assign "fields" values from the ICU data
+     *
+     * @param \ResourceBundle $localeData
+     */
+    private function assignFieldsValues(\ResourceBundle $localeData): void
+    {
+        /**
+         * Fields value in the current position has been added to ICU Data tables
+         * starting with ICU library version 51.1.
+         * Due to fact that we do not use these variables in templates, we do not initialize them for older versions
+         *
+         * @see https://github.com/unicode-org/icu/blob/release-50-2/icu4c/source/data/locales/en.txt
+         * @see https://github.com/unicode-org/icu/blob/release-51-2/icu4c/source/data/locales/en.txt
+         */
+        if ($localeData->get('fields')) {
+            $this->assign('today', $this->encoder->encode($localeData['fields']['day']['relative']['0']));
+            $this->assign('week', $this->encoder->encode($localeData['fields']['week']['dn']));
+        }
+    }
+}
```

**Comment by ihor-sviziev** [comment]
Hi @davidalger,
Seems like this issue was already resolved in the 2.4.0 release, related changes you can see here https://github.com/magento/magento2/commit/c4ca054a7dbf1454a32bb256314958bac1ebd952

Could you confirm that issue is already fixed?

**Comment by davidalger** [comment]
Yes, this was resolved.
